File: leetcode/src/main/java/TwoSortedArrayMedian.java
/**
 * There are two sorted arrays nums1 and nums2 of size m and n respectively.<p>
 * Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
 *
 * {@see https://leetcode.com/problems/median-of-two-sorted-arrays/#/description}
 *
 * @author solarknight.zph created on 17/4/8 下午8:19
 * @version 1.0
 */
public class TwoSortedArrayMedian {

  public static double findMedianSortedArrays(int[] nums1, int[] nums2) {
    int l1 = nums1.length, l2 = nums2.length;
    if (l1 > l2) {
      return findMedianSortedArrays(nums2, nums1);
    }

    if (l1 == 0) {
      return (nums2[(l2 - 1) / 2] + nums2[l2 / 2]) / 2.0;
    }

    int left = 0, right = 2 * l1;
    while (left <= right) {
      int mid1 = (left + right) / 2;
      int mid2 = l1 + l2 - mid1;

      int lv1 = mid1 == 0 ? Integer.MIN_VALUE : nums1[(mid1 - 1) / 2];
      int rv1 = mid1 == 2 * l1 ? Integer.MAX_VALUE : nums1[mid1 / 2];
      int lv2 = mid2 == 0 ? Integer.MIN_VALUE : nums2[(mid2 - 1) / 2];
      int rv2 = mid2 == 2 * l2 ? Integer.MAX_VALUE : nums2[mid2 / 2];

      if (lv1 <= rv2 && lv2 <= rv1) {
        return (Math.max(lv1, lv2) + Math.min(rv1, rv2)) / 2.0;
      } else if (lv1 > rv2) {
        right = mid1 - 1;
      } else {
        left = mid1 + 1;
      }
    }
    return -1;
  }


  public static void main(String[] args) {
    int[] array1, array2;
    array1 = new int[]{1, 3};
    array2 = new int[]{2};
    System.out.println(findMedianSortedArrays(array1, array2));

    array1 = new int[]{1, 2};
    array2 = new int[]{3, 4};
    System.out.println(findMedianSortedArrays(array1, array2));

    array1 = new int[]{6, 10};
    array2 = new int[]{1, 2, 3, 4, 5, 7, 8, 9};
    System.out.println(findMedianSortedArrays(array1, array2));
  }
}

File: leetcode/src/main/golang/two_sorted_array_median.go
import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    if len(nums1) > len(nums2) {
        return findMedianSortedArrays(nums2, nums1)
    }

    len1, len2 := len(nums1), len(nums2)
    if len1 == 0 {
        return float64(nums2[(len2 - 1) / 2] + nums2[len2 / 2]) / 2
    }

    for l, r := 0, 2 * len1; l <= r; {
        mid1 := (l + r) / 2
        mid2 := len1 + len2 - mid1
        l1, r1 := cutArray(nums1, mid1)
        l2, r2 := cutArray(nums2, mid2)

        if l1 <= r2 && l2 <= r1 {
            return float64(ifOrFail(l1 > l2, l1, l2) + ifOrFail(r1 < r2, r1, r2)) / 2
        } else if l1 > r2 {
            r = mid1 - 1
        } else {
            l = mid1 + 1
        }
    }
    return -1
}

func cutArray(nums []int, cut int) (l, r int) {
    if cut == 0 {
        l = math.MinInt32
        r = nums[cut / 2]
    } else if cut == 2 * len(nums) {
        l = nums[(cut - 1) / 2]
        r = math.MaxInt32
    } else {
        l = nums[(cut - 1) / 2]
        r = nums[cut / 2]
    }
    return
}

func ifOrFail(cond bool, val1 int, val2 int) int {
    if cond {
        return val1
    } else {
        return val2
    }
}